### DISCUSSION TITLE
Authentication fail in plain mode

### DISCUSSION THREAD

**Discussion by wallace-hu** [emqx/emqx#5689] [Q&A]
Hi  I am using emqx enterprise version 4.3.1, and  testing the mnesia authentication function. 
 
When i config the mnesia plugin with below setting:

```
## Password hash.
##
## Value: plain | md5 | sha | sha256 | sha512
auth.mnesia.password_hash = sha256

##--------------------------------------------------------------------
## ClientId Authentication
##--------------------------------------------------------------------

## Examples
auth.client.1.clientid = test_id
auth.client.1.password = test_passwd
```

Then using the MQTT.fx with plain clientid and password  connecting the EMQX ,It succeed.

But if i modify the auth.mnesia.password_hash = plain.  then use the same plain clientid and password,  I got failed connection.

So my question is that what is the difference between "plain" and "sha256" mode. In my understanding,  If using "plain" mode, I should use the plain clientid <test_id> and password<test_passwd>  in the connection.

While if using "sha256", I should use  clientid<test_id> and the sha256 encrypted  password<a89293c100fd6f945750883727a7bcdb98bfc4020b231285aca2127d41654627>.

It seems i got wrong understanding,  What should i do if i want to take the sha256 encrypted text when connecting to the EMQ?

**Comment by k32** [comment]
Hello,

You need to re-create all users after changing the hashing method. It is impossible to change the existing hashes in the database because hashing is one-way.

**Reply by ycsin** [reply]
How to recreate all users?

**Reply by wallace-hu** [reply]
> Hello,
> 
> You need to re-create all users after changing the hashing method. It is impossible to change the existing hashes in the database because hashing is one-way.

Thks for the reply,  I am just confuesd that when i use "sha256" passwd hash mode,  I  can still connect the emqx with plain passwd "user_passwd" as below setting.

auth.mnesia.password_hash = sha256

auth.client.1.clientid = remote_mosquitto_client
auth.client.1.password = user_passwd


If i use below setting:
auth.mnesia.password_hash = plain

auth.client.1.clientid = remote_mosquitto_client
auth.client.1.password = user_passwd

what should i use when conect to the emqx?

**Reply by k32** [reply]
The hashing method doesn't affect the credentials that the client should send to the broker. It only affects how the client secret is stored internally in EMQ X.

**Reply by philippdzm** [reply]
> How to recreate all users?

via the cli

`emqx_ctl user delete id`

But this does not solve the problem. My user from my config gets re-created after restarting the mnesia plugin. Unfortunately sha512 does not work. I mentioned this problem here https://github.com/emqx/emqx-auth-mnesia/issues/19#issuecomment-1092791062
